src/rust/engine/nailgun/src/server.rs

```diff
@@ -0,0 +1,355 @@
+          result
+        }
+      });
+      connection_started.notified().await;
+    };
+
+    // Before exiting, acquire write access on the ongoing_connections lock to prove that all
+    // connections have completed.
+    debug!("Server waiting for connections to complete...");
+    let _ = ongoing_connections.write().await;
+    debug!("All connections completed.");
+    result
+  }
+
+  ///
+  /// The port that the server is listening on.
+  ///
+  pub fn port(&self) -> u16 {
+    self.port
+  }
+
+  ///
+  /// Returns a Future that will shut down the server by:
+  /// 1. stopping accepting new connections
+  /// 2. waiting for all ongoing connections to have completed
+  ///
+  pub async fn shutdown(self) -> Result<(), String> {
+    // If we fail to send the exit signal, it's because the task is already shut down.
+    let _ = self.exit_sender.send(());
+    self
+      .exited_receiver
+      .await
+      .map_err(|_| "Server exited uncleanly.".to_owned())?
+  }
+}
+
+pub struct RawFdExecution {
+  pub cmd: execution::Command,
+  pub stdin_fd: RawFd,
+  pub stdout_fd: RawFd,
+  pub stderr_fd: RawFd,
+}
+
+///
+/// Implementations of this trait should _not_ close the input file handles, and should let the
+/// caller do so instead.
+///
+pub trait RawFdRunner: Fn(RawFdExecution) -> ExitCode + Send + Sync {}
+
+impl<T: Fn(RawFdExecution) -> ExitCode + Send + Sync> RawFdRunner for T {}
+
+///
+/// A Nail implementation that proxies stdio to file handles that can be consumed by the given
+/// callback function.
+///
+/// If any of stdio is a tty (detected via environment variables special cased in our nailgun
+/// client), we ignore the input fds and open new fds directly to the tty path, which happens to
+/// be addressable as a file in OSX and Linux. In that case, there is no middle-man, and we
+/// ignore the fds opened by the nailgun server for data sent via the protocol.
+///
+#[derive(Clone)]
+struct RawFdNail {
+  executor: Executor,
+  runner: Arc<dyn RawFdRunner>,
+}
+
+impl Nail for RawFdNail {
+  fn spawn(
+    &self,
+    cmd: execution::Command,
+    input_stream: mpsc::Receiver<ChildInput>,
+  ) -> Result<Child, io::Error> {
+    let env = cmd.env.iter().cloned().collect::<HashMap<_, _>>();
+
+    // Handle stdin.
+    let (stdin_handle, stdin_sink) = Self::input(Self::ttypath_from_env(&env, 0))?;
+    let accepts_stdin = if let Some(mut stdin_sink) = stdin_sink {
+      let mut input_stream = input_stream.filter_map(|child_input| {
+        Box::pin(async move {
+          match child_input {
+            ChildInput::Stdin(bytes) => Some(Ok(bytes)),
+            ChildInput::StdinEOF => None,
+          }
+        })
+      });
+      let _join = self.executor.spawn(async move {
+        stdin_sink.send_all(&mut input_stream).map(|_| ()).await;
+      });
+      true
+    } else {
+      // Stdin will be handled directly by the TTY.
+      false
+    };
+
+    // And stdout/stderr.
```

Comment:
```suggestion
    // Now handle stdout/stderr.
```

Reply:
```suggestion
    // (2) Handle stdout/stderr.
```
alternative suggestion